core/src/main/scala/kafka/log/Log.scala

```diff
@@ -2390,16 +2390,23 @@ object Log extends Logging {
                                       producerStateManager: ProducerStateManager,
                                       logPrefix: String): Unit = {
     segmentsToDelete.foreach(_.changeFileSuffixes("", Log.DeletedFileSuffix))
+    val snapshotsToDelete = if (deleteProducerStateSnapshots)
+      segmentsToDelete.flatMap { segment =>
+        producerStateManager.removeAndMarkSnapshotForDeletion(segment.baseOffset)}
+    else
+      Seq()
 
     def deleteSegments(): Unit = {
       info(s"${logPrefix}Deleting segment files ${segmentsToDelete.mkString(",")}")
       val parentDir = dir.getParent
       maybeHandleIOException(logDirFailureChannel, parentDir, s"Error while deleting segments for $topicPartition in dir $parentDir") {
         segmentsToDelete.foreach { segment =>
           segment.deleteIfExists()
-          if (deleteProducerStateSnapshots)
-            producerStateManager.removeAndDeleteSnapshot(segment.baseOffset)
         }
+        snapshotsToDelete.foreach( snapshot => {
+          snapshot.deleteIfExists()
```

Comment:
Could we add some logging similar to the one in `LogSegment`?  https://github.com/apache/kafka/blob/3fb836f507d35e2a1ab39df57edca671ffc66073/core/src/main/scala/kafka/log/LogSegment.scala#L612-L615